### DISCUSSION TITLE
How to generate mixed grains with large size differences?

### DISCUSSION THREAD

**Discussion by stenpthcurry30** [neperfepx/neper#503] [Q&A] [answered]
Thanks for your great  tools.  @rquey , Could you explain or make a tutorial on how to generate mixed grains with large size  difference.
I want to use neper to generate the mixed grain with large size differences in 2D or 3D, it is similar to the structure of a large grain surrounded by a ring of small grains in the recrystallization process.
  To solve this problem, I thought I could control the grain size distribution and customize the seed distribution to solve this problem, but it never worked, the code is as follows：
```
neper -T -n from_morpho -dim 2  
-morpho "diameq:custom(mydisturb.txt),1-sphericity:lognormal(0.145,0.03)" 
-morphooptiini "coo:msfile(coo2.txt)" 
-domain "square(1,1)" 

mydisturb.txt:
0.09 500
0.3 2.0
0.6 1.0

coo2.txt:
0.5	0.5
0.374540119	0.020584494
0.374540119	0.969909852
0.950714306	0.611852895
0.950714306	0.139493861
0.950714306	0.292144649
0.950714306	0.366361843
0.950714306	0.456069984
0.950714306	0.785175961
0.950714306	0.199673782
0.950714306	0.514234438
0.950714306	0.592414569
0.950714306	0.046450413
0.731993942	0.170524124
0.731993942	0.065051593
0.731993942	0.948885537
0.731993942	0.965632033
0.731993942	0.097672114
0.598658484	0.034388521
0.598658484	0.909320402
0.15601864	0.969584628
0.15601864	0.775132823
0.15601864	0.939498942
0.15601864	0.89482735
0.15601864	0.921874235
0.15601864	0.088492502
0.15601864	0.195982862
0.15601864	0.045227289
0.15599452	0.271349032
0.15599452	0.828737509
0.15599452	0.28093451
0.15599452	0.140924225
0.15599452	0.802196981
0.15599452	0.074550644
0.15599452	0.986886937
0.058083612	0.772244769
0.058083612	0.198715682
0.058083612	0.005522117
0.058083612	0.815461428
0.058083612	0.706857344
0.058083612	0.729007168
0.058083612	0.771270347
0.058083612	0.074044652
0.058083612	0.358465729
0.058083612	0.11586906
0.866176146	0.863103426
0.866176146	0.330898025
0.866176146	0.06355835
0.866176146	0.310982322
0.866176146	0.325183322
0.866176146	0.729606178
0.866176146	0.887212743
0.601115012	0.025419127
0.601115012	0.107891427
0.708072578	0.031429186
0.708072578	0.907566474
0.708072578	0.07697991
```

The goal is to generate some small grains around a  large grain in the center,but it always ends in failure.Please point out the error of my code and provide some suggestions.

By the way , the coo2.txt is generated by python

**Comment by stenpthcurry30** [comment]
Hi , @rquey 
I have controlled the seeds to control the size of the grains. But the result isn‘t satisfactory. The codes are:
```
neper -T -n 5169 -morphooptiini "coo:msfile(coo.txt)" -morpho voronoi -dim 2 -domain "square:(1,1)"
```
By the way, I use python to generate the 5169 seeds. And the figure is:
![img](https://user-images.githubusercontent.com/95471485/177041951-e3734b39-64c0-4d19-9f74-78d755a31bb7.png)

It's so ugly.😂 I want to know whether it have the way to control grain size more accurately. I want the grain in the middle to be larger.I have controlled it by mean diameq. But the result is so unbelievable. 
![1](https://user-images.githubusercontent.com/95471485/177042173-c07bf1f9-ce96-41b2-9e59-beeec1cf957b.png)
![img1](https://user-images.githubusercontent.com/95471485/177042191-dbca707b-1649-40f8-82db-08152fcfe7f8.png)
Could you give me some advice to control the grain  size. @rquey 
And the coo.txt is here  [coo.txt](https://github.com/neperfepx/neper/files/9034863/coo.txt)
Wish you a happy life.

**Comment by rquey** [comment, marked as answer]
By using `-morpho "diameq:custom(mydisturb.txt)"` with
```
mydisturb.txt:
0.09 500
0.3 2.0
0.6 1.0
```
you are interpolating the grain size data between 0.09, 0.3 and 0.6, which is not what you want. (Plus, Neper expects equally spaced x values.)

To specify two (or more) grain sizes, provide a bimodal (or multimodal) size distribution, as in `-morpho "diameq:1*dirac(0.6)+1000*lognormal(0.09,0.02)"` - when the grain sizes are very different from each other, `-morphooptiini coo:packing` helps:
```
neper -T -n from_morpho -morpho "diameq:1*dirac(0.6)+100*lognormal(0.09,0.02),1-sphericity:lognormal(0.145,0.03)" -morphooptiini coo:packing,weight:radeq -morphooptistop itermax=1 -dim 2 -o d
neper -V d.tess -print img
```
![image](https://user-images.githubusercontent.com/2659376/177106641-7701da5f-800e-4da1-aa36-7ccbe25c8083.png)

However, this gives you no control on the position of the large grain...

Coming back to your experiments, the problem is that you are not assigning weights to the seeds.  As you used a "rejection distance" from the central seed, you should set the seed weight to that distance.

Actually, you can even start simpler and play only with the weights:
```
# Generate 1000 random positions
neper -T -n 1000 -dim 2 -statseed x,y

# Override 1st position with "0.5 0.5" (center)
sed -e "1c 0.5 0.5" n1000-id1.stseed > coos

# Set weights = 0.3 (radius) to first seed and 0 to others
echo "0.3" > weights
printf '0\n%.0s' {2..1000} >> weights

# Generate again
neper -T -n 1000 -dim 2 -morphooptiini "coo:file(coos),weight:file(weights)" -o d

# Visualize
neper -V d.tess -print img
```
![image](https://user-images.githubusercontent.com/2659376/177110633-9d0a8426-63f1-420f-b6ee-fb09bf2ce113.png)

This works better, bit it will generally generate small cells on the ring of the big grain.

This can be solved by setting a rejection area:

```
# Generate 10000 random positions
neper -T -n 10000 -dim 2 -statseed x,y

# Set 1st position to "0.5 0.5" (center)
echo "0.5 0.5" > coos

# Set other positions to random positions outside of the rejection area
awk '{if (($1-0.5)**2+($2-0.5)**2>0.3**2) print}' n10000-id1.stseed | head -999 >> coos

# Set weights = 0.3 (radius) to first seed and 0.03 to others
echo "0.3" > weights
printf '0.03\n%.0s' {2..1000} >> weights

# Generate again
neper -T -n 1000 -dim 2 -morphooptiini "coo:file(coos),weight:file(weights)" -o d

# Visualize
neper -V d.tess -print img
```
![image](https://user-images.githubusercontent.com/2659376/177113264-3dcca643-9fb6-48b3-9139-ed0a8c853983.png)

**Reply by stenpthcurry30** [reply]
Thank you sir,  The code works well, but I wonder if I can make the middle circle closer to the real grain. You know, the real grain should be close to the shape of six deformation.I try to append the output grain size to it, but the result is always strange. @rquey
